### PR TITLE
[TUTORIAL] Improve opt_gemm tutorial

### PR DESCRIPTION
**Pull request by yuruofeifei**
1. Currently, the number of this tutorial is not consistent with the text and the method to calculate numpy runtime is not as expected.

2. Added more optimization to bring runtime from around 1.2s to around 0.3s.

3. Added IR section after each optimization.

@yidawang

### CODE REVIEW

**Comment by tqchen** [comment]
cc @were

**Comment by tqchen** [comment]
@ZihengJiang  can you contribute an example with more level of tiling and get to actual numpy perf?

**Comment by yidawang** [comment]
This is an example running on MacOS. However, I believe @ZihengJiang 's scheduling should apply in most cases. On MacOS, by default `numpy.dot` links to [Accelerate](https://developer.apple.com/documentation/accelerate). I would assume it achieves similar performance as OpenBlas and MKL. Lastly, adding multi-threading is necessary, and consequently, we should use `time.time()` instead of `time.clock()` to measure the time for `numpy.dot`. A better (more apple-to-apple) comparison is to use `timeit`, since the tvm-built function averages elapsed times across multiple runs.

**Comment by masahi** [comment]
@yuruofeifei can you check whether enabling avx2 makes a difference?

**Comment by tqchen** [comment]
going to merge this in first for now if no body objects

**Comment by yidawang** [comment]
Please give me some time to do one round of review. Haven’t carefully read it yet.

File: tutorials/optimize/opt_gemm.py [hunk 1]
```diff
@@ -25,8 +25,8 @@
 `repo <https://github.com/flame/how-to-optimize-gemm>`_. Some of them have been applied by TVM
 abstraction automatically, but some of them cannot be simply applied due to TVM constraints.
 
-All the experiment results mentioned below, are executed on 2013's 15' MacBook equiped with
-Intel i7-2760QM CPU. The cache line size should be 64 bytes for all the x86 CPU.
+All the experiment results mentioned below, are executed on 2015's 15' MacBook equiped with
+Intel i7-4770QH CPU. The cache line size should be 64 bytes for all the x86 CPU.
```
Comment (yidawang) [inline tutorials/optimize/opt_gemm.py]: i7-4770HQ?

File: tutorials/optimize/opt_gemm.py [hunk 2]
```diff
@@ -49,8 +49,12 @@
 # Random generated tensor for testing
 a = tvm.nd.array(numpy.random.rand(N, N).astype(dtype), tvm.cpu(0))
 b = tvm.nd.array(numpy.random.rand(N, N).astype(dtype), tvm.cpu(0))
+
+now = time.time()
 # The expected answer
 answer = numpy.dot(a.asnumpy(), b.asnumpy())
+np_runing_time = time.time() - now
+print("Numpy running time: %f" % np_runing_time)
```
Comment (yidawang) [inline tutorials/optimize/opt_gemm.py]: Probably we want to use `timeit` for a better time measurement

File: tutorials/optimize/opt_gemm.py [hunk 3]
```diff
@@ -65,49 +69,50 @@
...
 ################################################################################################
 # Blocking
 # --------
 # A important trick to enhance the cache hit rate is blocking --- data chunck will be computed
 # block by block. The memory access inside the block is a small neighbourhood which is with high
-# meomry locality. In this tutorial, I pick up 8, a relatively small value (8 ints < 64 bytes),
-# as the blocking size.
-#
+# memory locality. In this tutorial, I picked up 32 as the blocking factor. So the block will
+# fill 32 * 32 * sizeof(int) bytes in the cache whose total size is 32KB (L1 data cache)
```
Comment (yidawang) [inline tutorials/optimize/opt_gemm.py]: To be precise, `sizeof(float)`. Also, clearly specify the total size 4KB, which can be easily fit into the L1 cache.

File: tutorials/optimize/opt_gemm.py [hunk 4]
```diff
@@ -65,49 +69,50 @@
...
 
-bn = 8
+bn = 32
+s = tvm.create_schedule(C.op)
 # Blocking by loop tiling
-yo, xo, yi, xi = s[C].tile(C.op.axis[1], C.op.axis[0], bn, bn)
+xo, yo, xi, yi = s[C].tile(C.op.axis[0], C.op.axis[1], bn, bn)
 # Hoist reduction domain outside the blocking loop
-s[C].reorder(yo, xo, k, yi, xi)
+s[C].reorder(xo, yo, k, xi, yi)
```
Comment (yidawang) [inline tutorials/optimize/opt_gemm.py]: I don't quite understand why the orders should be changed from y-x to x-y, can you elaborate? The default data layout should be row-major, which makes it better for the x axis to be outside the y axis?

File: tutorials/optimize/opt_gemm.py (hunk 1 above)
Comment (yidawang) [reply tutorials/optimize/opt_gemm.py]: Also, in the beginning, we may want to highlight the final improvement we made based on the naive code. We should also emphasize that with the help of TVM, it was achieved by only a few lines of Python code.

File: tutorials/optimize/opt_gemm.py [hunk 5]
```diff
@@ -136,30 +141,115 @@
...
+
+print(tvm.lower(s, [A, B, C], simple_mode=True))
+
+###################################################################################################
+# Loop Permutation
+# -------------
+# If we look at the above IR, we can see the inner loop row data is vectorized and
+# B is transformed into PackedB. The traversal of PackedB is sequential now.
+# So we will look at the access pattern of A. In current schedule, A is accessed column by column
+# which is not cache friendly. If we change the nested loop order of k and inner row index,
```
Comment (yidawang) [inline tutorials/optimize/opt_gemm.py]: Specify the axis corresponding to the column of A, perhaps `yi`?

File: tutorials/optimize/opt_gemm.py [hunk 6]
```diff
@@ -136,30 +141,115 @@
...
+opt5_time = evaluator(a, b, c).mean
+print('Opt5: %f' % opt5_time)
+
 ##################################################################################################
 # Summary
 # -------
-# After applying three main tricks, we can achieve almost 90% performance of numpy.
-# Further observation is required to catch up with the performance of numpy.
+# After applying the above simple optimizations with only 9 lines of code,
+# our generated code can achieve 30% of numpy performance with Apple implemented BLAS.
```
Comment (yidawang) [inline tutorials/optimize/opt_gemm.py]: Finally, please refer to a similar implementation from @javelinjs for more insight. https://github.com/javelinjs/topi-intel/blob/master/opt_gemm.py

File: tutorials/optimize/opt_gemm.py (hunk 1 above)
Comment (yidawang) [reply tutorials/optimize/opt_gemm.py]: And perhaps add your name to the author list.

File: tutorials/optimize/opt_gemm.py (hunk 6 above)
Comment (yzhliu) [reply tutorials/optimize/opt_gemm.py]: I have no idea why cache_write for C actually hurts the performance in my schedule, if anyone can give some hints.
BTW, I just changed my github username so as to be more easily mentioned.

File: tutorials/optimize/opt_gemm.py (hunk 1 above)
Comment (yuruofeifei) [reply tutorials/optimize/opt_gemm.py]: Changed to i7-4770HQ

File: tutorials/optimize/opt_gemm.py (hunk 2 above)
Comment (yuruofeifei) [reply tutorials/optimize/opt_gemm.py]: Changed

File: tutorials/optimize/opt_gemm.py (hunk 3 above)
Comment (yuruofeifei) [reply tutorials/optimize/opt_gemm.py]: changed

File: tutorials/optimize/opt_gemm.py (hunk 4 above)
Comment (yuruofeifei) [reply tutorials/optimize/opt_gemm.py]: The generated IR is using x as row index and y as column index, I changed name to make it align with generated IR. I can revert it if that makes more sense to you.

File: tutorials/optimize/opt_gemm.py (hunk 6 above)
Comment (yuruofeifei) [reply tutorials/optimize/opt_gemm.py]: yzhliu's code is very insightfuly, I probably will have time later this week to investigate more.

**Comment by yidawang** [comment]
Please merge the latest changes from `master` before merging. Thanks!

**Pull request merged**